sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala

```diff
@@ -1310,6 +1310,14 @@ class PlannerSuite extends SharedSparkSession with AdaptiveSparkPlanHelper {
     assert(topKs.size == 1)
     assert(sorts.isEmpty)
   }
+
+  test("SPARK-39915: Dataset.repartition(N) may not create N partitions") {
```

Comment:
This is not a planner bug... We can probably add an end-to-end test in `DataFrameSuite`